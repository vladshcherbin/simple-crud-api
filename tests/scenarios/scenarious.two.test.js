import crypto from 'crypto'
// eslint-disable-next-line import/no-extraneous-dependencies
import { jest } from '@jest/globals'
import request from 'supertest'

const samplePersons = [
  {
    id: '02d2de0d-28f6-4d9a-b2cd-d9725b8d16c1',
    name: 'Alex',
    age: 18,
    hobbies: ['tennis']
  },
  {
    id: '75bae368-133e-4896-9655-19ddf027ad14',
    name: 'Sam',
    age: 6,
    hobbies: []
  },
  {
    id: 'b73a288b-7a09-4fa6-b988-2b14235e1e95',
    name: 'Susan',
    age: 41,
    hobbies: ['dancing', 'hiking']
  }
]
const secondPersonId = samplePersons[1].id
const mockedRandomUUID = jest.fn(() => crypto.randomUUID())

jest.unstable_mockModule('crypto', () => ({
  default: { ...crypto, randomUUID: mockedRandomUUID }
}))

jest.unstable_mockModule('../../src/data/persons', () => ({
  default: samplePersons
}))

const { default: server } = await import('../../src/server')

describe('Scenario two', () => {
  test('Return person list with 3 persons', async () => {
    const response = await request(server.instance()).get('/person')

    expect(response.status).toBe(200)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Delete second person', async () => {
    const response = await request(server.instance()).delete(`/person/${secondPersonId}`)

    expect(response.status).toBe(204)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Return 404 error for deleted person', async () => {
    const response = await request(server.instance()).get(`/person/${secondPersonId}`)

    expect(response.status).toBe(404)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Return 400 trying to update first person with invalid data', async () => {
    const response = await request(server.instance()).put(`/person/${samplePersons[0].id}`).send({
      age: 19
    })

    expect(response.status).toBe(400)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Update first person with valid data', async () => {
    const response = await request(server.instance()).put(`/person/${samplePersons[0].id}`).send({
      name: 'Alex',
      age: 19,
      hobbies: ['tennis']
    })

    expect(response.status).toBe(200)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })

  test('Return person list with 2 persons', async () => {
    const response = await request(server.instance()).get('/person')

    expect(response.status).toBe(200)
    expect(response.type).toBe('application/json')
    expect(response.body).toMatchSnapshot()
  })
})
