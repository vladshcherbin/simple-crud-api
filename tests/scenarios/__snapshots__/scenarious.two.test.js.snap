// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scenario two Delete second person 1`] = `""`;

exports[`Scenario two Return 400 trying to update first person with invalid data 1`] = `
Object {
  "errors": Object {
    "hobbies": "\\"hobbies\\" value is missing",
    "name": "\\"name\\" value is missing",
  },
  "message": "Validation error",
}
`;

exports[`Scenario two Return 404 error for deleted person 1`] = `
Object {
  "message": "Person not found",
}
`;

exports[`Scenario two Return person list with 2 persons 1`] = `
Array [
  Object {
    "age": 19,
    "hobbies": Array [
      "tennis",
    ],
    "id": "02d2de0d-28f6-4d9a-b2cd-d9725b8d16c1",
    "name": "Alex",
  },
  Object {
    "age": 41,
    "hobbies": Array [
      "dancing",
      "hiking",
    ],
    "id": "b73a288b-7a09-4fa6-b988-2b14235e1e95",
    "name": "Susan",
  },
]
`;

exports[`Scenario two Return person list with 3 persons 1`] = `
Array [
  Object {
    "age": 18,
    "hobbies": Array [
      "tennis",
    ],
    "id": "02d2de0d-28f6-4d9a-b2cd-d9725b8d16c1",
    "name": "Alex",
  },
  Object {
    "age": 6,
    "hobbies": Array [],
    "id": "75bae368-133e-4896-9655-19ddf027ad14",
    "name": "Sam",
  },
  Object {
    "age": 41,
    "hobbies": Array [
      "dancing",
      "hiking",
    ],
    "id": "b73a288b-7a09-4fa6-b988-2b14235e1e95",
    "name": "Susan",
  },
]
`;

exports[`Scenario two Update first person with valid data 1`] = `
Object {
  "age": 19,
  "hobbies": Array [
    "tennis",
  ],
  "id": "02d2de0d-28f6-4d9a-b2cd-d9725b8d16c1",
  "name": "Alex",
}
`;
